### PR TITLE
#95 Multi-order Submission

### LINKED ISSUE #95: Multi-order Submission for Regular Accounts
"Order placement function accepts arrays of order parameters and throws several events."

Currently users must submit their orders individually paying gas fees for each order transaction. It would be nice to be able to submit multiple orders in a single transaction (this would reduce the gas costs for multiple orders and also enable us to unit test what happens when an order batch is filled before the batch closes. In this case, several events or type `SellOrder` would be emitted.

### PR DESCRIPTION
**Pull request by bh2smith**
Closes #95

Test Plan - Unit tests.

### CODE REVIEW

**Review by josojo — APPROVED** [review]
I like the PR because it is very readable and reuses a lot of existing code. 
However, it could be written more gas efficiently. Just mentioning it here.
 I guess readability is currently more important, gas improvements can still be done later.

**Comment by bh2smith** [comment]
@josojo 

> However, it could be written more gas efficiently

Do you have a suggestion how it could be more gas efficient? Notice that all the `require` statements should be applied to each order (since they could have a wrong `tokenId` in any one of the orders. I have considered your suggestion, but didn't see too many places when gains could be made.

**Comment by josojo** [comment]
At least the following three steps could be omitted:
1. onlyRegisteredAccounts() modifier
2. getting the accountId, its always the same
3. Throwing only one log statement with arrays instead of several ones.
4. Not sure, whether the compiler is smart enough to not store the hash update with a sstore, but just keeping it in memory.

But still, I do not think that it is worth the effort.

**Comment by bh2smith** [comment]
@josojo  I completely agree with the `accoutId` retrieval, but don't necessarily like the single event (because we would then have to propagate all these changes to the listener as well). Actually this could be a really good refactoring improvement to give to a new employee during "onboarding". I am not sure what is the problem with `onlyRegistered`. You mean the fact that it is being called for every single order instead of just once? Yes, I agree this is also less efficient and should probably be changed. 

Summary, 

1. Should probably be changed. Good task for new contributor?
2. Should also be changed (change could be made in the same place as Item 1)
3. Would require more changes in the Listener as well [Really good onboarding task for new contributor] Should we create an issue about this  and label it "good first issue"?

**Comment by bh2smith** [comment]
Just waiting on @fleupold

**Comment by fleupold** [comment]
What was our consensus on using packed blobs instead of multiple arrays which have to implicitly aligned? 
I believe in #97 @josojo offered to change this immediately. Maybe we can include it here so given that the PR is quite small.

Otherwise I agree with 1), 2), and 4). Making this more gas efficient could be a good "ToDo" task. I do prefer reusing the same event as explained in #101

**Comment by bh2smith** [comment]
I was under the impression that Blob order data would be changed after these draft PRs were accepted. I don't consider it such a "small change" and I didn't cant see any part of #97 regarding this immediate change.

**Comment by fleupold** [comment]
Sure, however looking at the current diff most of the code written here (especially the unit tests) will be affected by that change. So not sure if it's worth committing this before. But if you want to feel free.

Yet, I would consider this task part of the original issue.

File: .solhintignore
Comment (fleupold) [inline .solhintignore]: unrelated changes. Can you base this diff off the one that introduced those to make reviewing easier?

File: contracts/SnappAuction.sol
Comment (fleupold) [inline contracts/SnappAuction.sol]: When using ECRecover we need to attach a signature to each order, which is 65 bytes of extra data. This will allow us to recover the ethereum public address which we can then use to look up the accountId. Check https://github.com/protofire/zeppelin-solidity/blob/master/contracts/ECRecovery.sol for more info on how ecrecover works.

File: contracts/SnappAuction.sol
Comment (fleupold) [inline contracts/SnappAuction.sol]: For now, I'd just make sure it's msg.sender until we introduce proper signature checking.

File: contracts/SnappAuction.sol
Comment (fleupold) [inline contracts/SnappAuction.sol]: I think you can still explicitly cast down to uint96

File: contracts/SnappAuction.sol
Comment (fleupold) [inline contracts/SnappAuction.sol]: Why 27? Isn't it 2 for accountId, 1 for buy + 1 for sell token & 12 for buy + 12 for sell token? That would be 28 bytes, wouldn't it?

File: contracts/SnappAuction.sol
Comment (fleupold) [inline contracts/SnappAuction.sol]: Can we for now just unpack the order and call the original `postSellOrder` function n times? Otherwise let's refactor postSellOrder to call into `placeMultiSellOrder`. At the end I'd really just like to maintain one unreserved sell order event in the listener.

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol]: Yes it should be 28... Not sure where I lost one.

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol]: No, because I can't get `uint128` from the unpacking function.

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol]: I was not able to cast down to 128, I don't see how I will be able to cast down even more.

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol]: Do you mean require that we should fetch the accountId of msg.sender and ensure that they are registered? Meaning use the modifier `onlyRegistered()`

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol]: Thanks, although I was aware of how ECRecover works. I suppose the 16 bits is what threw you off.

File: .solhintignore
Comment (bh2smith) [reply .solhintignore]: This branch assumes whitespace2 has been merged. I don't plan on reversing these.

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol]: Can you tell me how you are able to comment on multiple lines!

File: contracts/SnappAuction.sol
Comment (josojo) [reply contracts/SnappAuction.sol]: Yeah, it is easier to refactor `postSellOrder`, agreed.

Comment (bh2smith) [reply contracts/SnappAuction.sol]: Cool, but we have a gas issue to deal with first.
I have concluded that either 
- commenting `unpackAndVerifyOrderData(orderData);` or
- replacing `bytes32 nextAuctionHash = sha256(abi.encodePacked(auctions[auctionIndex].shaHash, orderData));` with `bytes32 nextAuctionHash = 0x0`

would both work...

```js
function placeMultiSellOrder(bytes memory packedOrders) public onlyRegistered() {
    // Note that this could result failure of all orders if even one fails.
    require(packedOrders.length % 28 == 0, "Each order should be packed in 28 bytes!");

    for (uint i = 0; i < packedOrders.length / 28; i++) {
        bytes memory orderData = packedOrders.slice(28*i, 28*(i+1));
        unpackAndVerifyOrderData(orderData);
        orderBatchUpdate();

        bytes32 nextAuctionHash = sha256(abi.encodePacked(auctions[auctionIndex].shaHash, orderData));
        auctions[auctionIndex].shaHash = nextAuctionHash;

        emit PackedSellOrder(auctionIndex, auctions[auctionIndex].size, orderData);
        // Only increment size after event (so it is emitted as an index)
        auctions[auctionIndex].size++;
    }
}

```

File: contracts/SnappAuction.sol
Comment (fleupold) [reply contracts/SnappAuction.sol]: Yes, the comment is weird. It sounded as if you want to recover something from 16 bits, where really you need 65 bytes of data and can only recover the main Ethereum address which still needs to be looked up. Can you adjust the comment?

File: contracts/SnappAuction.sol
Comment (fleupold) [reply contracts/SnappAuction.sol]: > Can you tell me how you are able to comment on multiple lines!

In solidity or in github? I don't think I can comment on multiple lines in github. In solidity using `/* ... */`

> Do you mean require that we should fetch the accountId of msg.sender and ensure ...

yes, and ensure that accountId is that value. In fact looking at the code this is what is happening becuase the other line is commented out. Is the comment still relevant?

File: contracts/SnappAuction.sol
Comment (fleupold) [reply contracts/SnappAuction.sol]: Does this not work:

```
uint foo = 1;
uint96 bar = uint96(foo);
```

?

File: contracts/SnappAuction.sol
Comment (fleupold) [reply contracts/SnappAuction.sol]: So is this related to the number of orders you are trying to place? As in does it work if you call `placeMultiSellOrder` with 0 bytes? At how many orders does it start failing? 

I guess we have to do some debugging to identify what is costing how much gas. You can use `it.only` and `truffle test -- gas` to effectively narrow down which call is causing how much gas.

File: .solhintignore
Comment (fleupold) [reply .solhintignore]: Sure, in the future you could base your PR on the one that it assumes will have been merged before. Otherwise it's really hard for your reviewers to see which changes are relevant to this PR.

**Comment by fleupold** [comment]
I believe that the out of gas error might come from the fact that our smart contract byte code is exceeding 45kb (https://ethereum.stackexchange.com/questions/7436/large-contract-bytecode-45k-causes-out-of-gas-on-deploy) now.

This would explain why making the source code somewhat shorter does not incur the error anymore.

You can check the length of the `bytecode` field in `build/SnappAuction.json` after callign `truffle compile`

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol]:
```suggestion
        // TODO - use ECRecover from signature contained in first 65 bytes of packedOrder
```

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol]: Unfortunately not exactly. I can confirm that 

```
uint _c = 2;
uint24 c = uint24(_c);
```
Does work, but that
```
uint _c = BytesLib.toUint(_data, 1 + 2);
uint24 c = uint24(_c);
```

does not. Actually, now I am beginning to wonder if this is an issue with `BytesLib.toUint`

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol]: > So is this related to the number of orders you are trying to place? As in does it work if you call placeMultiSellOrder with 0 bytes? At how many orders does it start failing?

Unfortunately, the payload is not the problem here. This is a problem with contract size during migration and I haven't even started trying to unit test this with real orders yet.

We are already very well aware of the gas consumption of this contract on deployment. The following is the gas cost for deployment of the current version of the contract

```
   Deploying 'SnappAuction'
   ------------------------
   > transaction hash:    0x0026525ad588bd74b56e63ef6a43101f24f2b962bd1023221d2c9ad9a56d6f60
   > Blocks: 0            Seconds: 0
   > contract address:    0x9A2400BDeE35f01Deb6d6Cbdd24f0d2eAF9E92D5
   > account:             0xa0c455c7e343f2799E26f085bA8Ee671DDd2434A
   > balance:             99.69985876
   > gas used:            7010022
   > gas price:           20 gwei
   > value sent:          0 ETH
   > total cost:          0.14020044 ETH
```

**Comment by bh2smith** [comment]
> I believe that the out of gas error might come from the fact that our smart contract byte code is exceeding 45kb (https://ethereum.stackexchange.com/questions/7436/large-contract-bytecode-45k-causes-out-of-gas-on-deploy) now.

Yes, @fleupold I agree, looking forward to dealing with this one. We are saving 30% if compiled "with optimization", but this is arguably not a solution. Also, apparently the optimizer is not audited.

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol]: Im trying to see how you are able to comment on 4 lines of code here at the once?

Screenshot 2019-06-08 at 21 40 55

File: .solhintignore
Comment (bh2smith) [reply .solhintignore]: The problem I see with that is that I would either
1. have to look at those hideous lint errors in my editor while working on this (which I am not willing to do) or 
2. wait until the other branch is merged before submitting this at all. 

Let hope that there are no more gross whitespace problems in the future and this will no longer be a concern.

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol]: Yes, it turns out, that the bytesLib `toUint` function has a length requirement on the input bytes length as seen here

https://github.com/GNSPS/solidity-bytes-utils/blob/14ca2bd38f96f3b18c06c0cf13df1997a3831a78/contracts/BytesLib.sol#L337-L346

This means I will have to slice and pad the info and then cast down. Luckily, this will allow us to work with the desired 96 bits right off the bat, and no need to fiddle around with 128 checks.

File: contracts/SnappAuction.sol
Comment (fleupold) [inline contracts/SnappAuction.sol]: `createNewPendingBatchIfNecessary`? `orderBatchUpdate` sounds a bit vague and it's not really clear from the name what the difference between this method and `createNewPendingBatch` is.

File: contracts/SnappAuction.sol
Comment (fleupold) [inline contracts/SnappAuction.sol:177]: I'd leave the old interface for convenience and just call into the new method from within the old method (creating a batch of 1 item)

File: contracts/SnappAuction.sol
Comment (fleupold) [inline contracts/SnappAuction.sol:206]: Not necessarily needed for now, but we should probably note that we could make this more fas efficient by only writing to the struct when we actually need to (upon new batch creation or in the last round).

File: contracts/SnappAuction.sol
Comment (josojo) [inline contracts/SnappAuction.sol:200]: Why do we use an encodePacked instead of encode? Is there a risk if a sha hash starts with a zero?

**Comment by josojo** [comment]
Very good progress. 
Another test which is calculating the sha hash offline and compares it to the sha hash in solidity would give us more confidence that the assembly code is working. ( not sure, how much work that is and whether it is really needed)

**Comment by bh2smith** [comment]
@josojo I am confident that the assembly code is working since its essentially a generalized version of this (e.g. `toUint32`)

https://github.com/GNSPS/solidity-bytes-utils/blob/master/contracts/BytesLib.sol

Furthermore, I believe that this can be tested by checking the value of the amounts in the events that are emitted in a unit test.

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol:200]: I don't understand your question. We have always been using `encodePacked` and I was just continuing to do so.

Comment (josojo) [reply contracts/SnappAuction.sol:200]: Yeah, we have always been using encodePacked, 
I was wondering, whether there is a risk in case we are getting leading zeros in the sha hash and in the id. In this case the backend infrastructure would not be aware of where to truncate. I think it could cause problems and we might want to switch to encode only.
However, it might all work fine, as every string starts with 0x anyways and then there will not be any truncating.  But then we can use encode anyways, instead of encodePacked

Comment (bh2smith) [reply contracts/SnappAuction.sol:200]: Perhaps, this would be better in the design discussion. I would love to learn more about this and see some examples/potential points of failure.

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol]: Yea cool!

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol:177]: Unfortunately not. That was the reason for the large contract size that wouldn't migrate. Do you still think is should be left here?

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol:206]: Sorry, I don't understand what is "not needed here". Are you saying that we don't need to update the `orderBatch` hash here?

File: contracts/SnappAuction.sol
Comment (bh2smith) [reply contracts/SnappAuction.sol:177]: That is to say, if we leave the old interface, then the contract won't migrate without optimization enabled. Is this what we really want?

File: contracts/SnappAuction.sol
Comment (fleupold) [reply contracts/SnappAuction.sol:206]: Writing `nextAuctionHash` to storage (which costs 5k gas) only needs to happen in the very last iteration of the loop (or if we create a new pending batch).

Right now it's happening once per loop.

However, making that change is not that trivial, so I'd recommend just adding a TODO for now.

File: contracts/SnappAuction.sol
Comment (fleupold) [reply contracts/SnappAuction.sol:177]: I see. I believe we should mitigate the large contract size for now by enabling optimisation (or even increasing the block gas limit on ganache), since otherwise we will reach the limit in the next PR again.

I will make it my priority to refactor the SnappBase into a library to give us a bit more breathing space long term for deploying within the gas limit.

File: test/snapp_auction.js
Comment (fleupold) [inline test/snapp_auction.js]: Do we really want to commit this?

**Review by fleupold — APPROVED** [review]
Looks good to me. I'd put back the convenience interface and enable optimisations for now. I will try to fix the gas issue asap.

File: test/snapp_auction.js
Comment (bh2smith) [reply test/snapp_auction.js]: I guess not, but I really want to make it clear that this is a pending topic of concern. I will create an issue for this instead.

File: contracts/SnappAuction.sol
Comment (fleupold) [inline contracts/SnappAuction.sol:154]: I was hoping this function could just call the new function with a byte array containing the single order encoded (instead of duplicating the logic).

Comment (bh2smith) [reply contracts/SnappAuction.sol:154]: Yes, that would have been nice. Too bad this wasn't mentioned. What do you suggest I do now?

Comment (fleupold) [reply contracts/SnappAuction.sol:154]: I tried to mention it in https://github.com/gnosis/dex-contracts/pull/100#discussion_r292090784

> I'd leave the old interface for convenience and just call into the new method from within the old method (creating a batch of 1 item)

We can create a new PR with just that small improvement.

Comment (bh2smith) [reply contracts/SnappAuction.sol:154]: Yes very nice!

Comment (bh2smith) [reply contracts/SnappAuction.sol:154]: I would argue that this isn't going to work very well with the double application of `onlyRegistered` and the fact that we have to write an entirely different version of "encodeOrder" that make 26 bytes out of the order data.

From my side, this function would be better off dead.

Comment (fleupold) [reply contracts/SnappAuction.sol:154]: I'll take a crack at it if you don't mind.

Comment (bh2smith) [reply contracts/SnappAuction.sol:154]: Although this is still far from correct, this is the best I can do in terms of this suggestion
```
    function placeSellOrder(
        uint8 buyToken,
        uint8 sellToken,
        uint96 buyAmount,
        uint96 sellAmount
    ) public onlyRegistered() {
        uint orderInt = uint(buyToken) + (uint(sellToken) << 8) + (uint(sellAmount) << 16) + (uint(buyAmount) << 112);
        bytes memory orderBytes = new bytes(26);
        assembly { mstore(add(orderBytes, 26), orderInt) }
        placeSellOrders(orderBytes);
```

Comment (bh2smith) [reply contracts/SnappAuction.sol:154]: Yes, please @fleupold

**Pull request merged**
